Update approx current_flow betweenness to use k directly

mrecachinas opened the pull request:
Rough pass at a fix for #6188.

amcandio commented on networkx/algorithms/centrality/current_flow_betweenness.py:
```diff
@@ -119,10 +124,13 @@ def approximate_current_flow_betweenness_centrality(
     nb = (n - 1.0) * (n - 2.0)  # normalization factor
     cstar = n * (n - 1) / nb
     l = 1  # parameter in approximation, adjustable
```
This l is also odd. Should we make it a parameter?

rossbar reviewed (commented):
Thanks for putting this together @mrecachinas - seeing the changes to the test suite make me less enthused about the "get rid of epsilon" approach. Given this, I'm leaning more towards docstring updates as the best way to resolve #6188: essentially improving the description for both `epsilon` and `kmax` and mentioning the relationship (as you had originally proposed @mrecachinas !). WDYT?

Also on an unrelated note - I see the old normalization factor  (see L119) - pinging @eriknw for an extra set of eyes given the work to fix bc in #7908 . There's no `endpoints` in this case, but there is `k`!

dschult commented on networkx/algorithms/centrality/current_flow_betweenness.py:
```diff
@@ -29,14 +29,13 @@ def approximate_current_flow_betweenness_centrality(
     solver="full",
-    epsilon=0.5,
-    kmax=10000,
+    k=None,
     seed=None,
```
Let's make the new keyword arg keyword only.  And we should have `seed` keyword only also -- because it changes position in the function signature when we replace `epsilon` and `kmax` with `k`.  So, anyone using position args will need to change their code -- and they should make the call use keywords.

So, I think something like this should work:
```
*, 
k=None,
seed=None,
```
Also -- changing the argument structure will require that the random_state decorator get updated too. 
```python
@py_random_state("seed")
```

dschult commented:
I think this is different than the other BC function.  First and most importantly, this doesn't have the `endpoints` kwarg. Next, it was already set up to handle the number of sample pairs `k`. And finally, there wasn't any dependence on kmax or epsilon before. So there shouldn't be any normalization dependence on k now.

mrecachinas commented:
Happy to remove my most recent changes (saw them after @dschult's review). I reverted the change to `k`, updated `epsilon`'s description, and pulled `l` out to `sample_weight`, which seems appropriate since it's scaling the sampling process that yields `k`. 🤔

amcandio reviewed (approved):
Looks good to me. Many of the inputs such as n, sample size of epsilon could cause zero division errors. We might want to add some validations on the input so the error is clear for the user

mrecachinas commented:
Not sure why the label check is failing. Should it have been auto-added?

mrecachinas commented on networkx/algorithms/centrality/current_flow_betweenness.py:
```diff
@@ -118,8 +142,8 @@ def approximate_current_flow_betweenness_centrality(
+    k = sample_weight * int(np.ceil((cstar / epsilon) ** 2 * np.log(n)))
     if k > kmax:
         msg = f"Number random pairs k>kmax ({k}>{kmax}) "
         raise nx.NetworkXError(msg, "Increase kmax or epsilon")
```
Done in [44bc0ef](https://github.com/networkx/networkx/pull/8007/commits/44bc0efa44331987e4b71b945ffb790bd7919e3c)

rossbar commented on networkx/algorithms/centrality/current_flow_betweenness.py:
```diff
@@ -21,16 +21,18 @@
 @nx._dispatchable(edge_attrs="weight")
 def approximate_current_flow_betweenness_centrality(
     G,
+    *,
```
Unfortunately this breaks backward compatibility - only newly-added kwargs can be key-word only, not the existing ones.

rossbar replied:
The backward-compatible recipe would be to leave the ordering of the existing kwargs, then `*`, then the new kwarg, so:

```python
def approximate_current_flow_betweenness_centrality(
    G,    
    normalized=True,
    weight=None,
    dtype=float,
    solver="full",
    epsilon=0.5,
    kmax=10000,
    seed=None,
    *,
    sample_weight=1,
):
```

mrecachinas replied:
Done in [f1c72be](https://github.com/networkx/networkx/pull/8007/commits/f1c72be51455183445c27ae78984f7f29943acfd)

Schefflera-Arboricola reviewed (commented):
Thanks @mrecachinas ! Consider adding tests for the new `sample_weigh` parameter and the error messages.

rossbar commented on networkx/algorithms/centrality/current_flow_betweenness.py:
```diff
@@ -64,11 +66,18 @@ def approximate_current_flow_betweenness_centrality(
 
     epsilon: float
-        Absolute error tolerance.
+        Absolute error tolerance. Note that smaller values of epsilon lead to
+        higher numbers of sample pairs (`k`) and thus more computation time. The number
+        of sample pairs is approximately `(c/epsilon)^2 * log(n)` where `n` is the
+        number of nodes.
```
```suggestion
        Absolute error tolerance. Note that smaller values of `epsilon` lead to
        higher numbers of sample pairs (``k``) and thus more computation time. The number
        of sample pairs is approximately ``(c/epsilon)^2 * log(n)`` where ``n`` is the
        number of nodes.
```

rossbar commented on networkx/algorithms/centrality/current_flow_betweenness.py:
```diff
@@ -64,11 +66,18 @@ def approximate_current_flow_betweenness_centrality(
     kmax: int
        Maximum number of sample node pairs to use for approximation.
 
+    sample_weight: float (default=1)
```
It'd be good to make this change for the other parameters as well!

```suggestion
    sample_weight : float (default=1)
```

rossbar reviewed (approved):
LGTM, thanks @mrecachinas 

I left a couple final docstring formatting nits. Normally I'd just push them up myself but it looks like you have the "maintainers may edit this branch" button disabled. Sorry for the nits!

MridulS commented on networkx/algorithms/centrality/current_flow_betweenness.py:
```diff
@@ -102,12 +111,34 @@ def approximate_current_flow_betweenness_centrality(
+        return dict.fromkeys(G, 0.0)
+
+    if epsilon <= 0:
+        raise nx.NetworkXError(f"Epsilon must be positive. Got epsilon={epsilon}.")
+
+    if sample_weight <= 0:
+        raise nx.NetworkXError(
+            f"Sample weight must be positive. Got sample_weight={sample_weight}."
```
```suggestion
        raise nx.NetworkXError(f"Epsilon must be positive. Got {epsilon=}.")

    if sample_weight <= 0:
        raise nx.NetworkXError(
            f"Sample weight must be positive. Got {sample_weight=}."
```

dschult commented on networkx/algorithms/centrality/current_flow_betweenness.py:
```diff
@@ -102,12 +111,34 @@ def approximate_current_flow_betweenness_centrality(
+        raise nx.NetworkXError(f"Epsilon must be positive. Got {epsilon=}.")
+
+    if sample_weight <= 0:
+        raise nx.NetworkXError(
+            f"Sample weight must be positive. Got {sample_weight=}."
+        )
```
fix linting
```suggestion
        raise nx.NetworkXError(f"Sample weight must be positive. Got {sample_weight=}.")
```

dschult reviewed (approved):
This looks good. I've suggested a linting fix below. And I approve this PR once we get a green check. :)
